Franck/question implementation

MojoOno opened the pull request:
This commit contains functionality + manual endpoint testing for the QuestionController.

An Interface for the QuestionController has been implemented with a helper method for the update method in QuestionController for a more readable method.
The methods have been refactored from the old HotelController and manually tested successfully for failure and success.

Endpoint for ALGEBRA has been outcommented and the same goes for the method inside QuestionController "Handling" that endpoint. It has not been deleted as i want confirmation before doing that.

The demo.http has been refactored to fit our documentation and attributes in Question entity and DTO

MojoOno committed:
Out commented endpoint for category

MojoOno committed:
Removed constructor with CrudDAO as input.
Added proper error handling for getAll method in order for it to fail properly.

MojoOno committed:
Added proper error handling for getById method in order for it to fail properly.

MojoOno committed:
Refactored the create method to fit Question and tested that it failed and succeded succesfully

Added proper error handling

MojoOno committed:
added fetchType.EAGER to the ManyToMany relations to assignments as it have me an error about fetching lazily

This made the getAll() succeed and fail successfully

MojoOno committed:
Added setters to Question

MojoOno committed:
Made an Interface for QuestionController as i needed to implement the method updateFieldIfNotNull()

MojoOno committed:
Added status code 201 to create()
Added functionality to the update()
Refactored to question in delete()

Added a helper method updateFieldIfNotNull() that is used to accept the updated value in update()

MojoOno committed:
- Removed unused imports
- Added status code 200
- Cleaned up getById() + added status code 200
- Removed logging from create() + added status code 201 + added error handling but is lacking the helper method to precise our error message.
- Removed logging from update method + added status code 200
- Cleaned up delete() + added exception handling.

MojoOno committed:
- Removed the unique condition for questionNumber
- Added nullable = false to questionText

MojoOno committed:
- Updated demo.http with functioning endpoint test

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -25,44 +25,37 @@ public QuestionController(EntityManagerFactory emf)
-            //long id = Long.parseLong(ctx.pathParam("id"));
+        try
+        {
             long id = ctx.pathParamAsClass("id", Long.class)
```
change to Integer.class

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -25,44 +25,37 @@ public QuestionController(EntityManagerFactory emf)
                     .getOrThrow((validator) -> new BadRequestResponse("Invalid id"));
-            HotelDTO foundEntity = new HotelDTO(dao.getById(Hotel.class, id));
-            ctx.json(foundEntity);
+            QuestionDTO foundEntity = new QuestionDTO(dao.getById(Question.class, id));
```
variable name change to foundQuestion

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -25,44 +25,37 @@ public QuestionController(EntityManagerFactory emf)
             long id = ctx.pathParamAsClass("id", Long.class)
-                    .check(i -> i>0, "id must be at least 0")
+                    .check(i -> i > 0, "id must be at least 0")
                     .getOrThrow((validator) -> new BadRequestResponse("Invalid id"));
```
check om vi få en String output = ctx.queryParam("output") med ind, og hvis den har værdien "full", så skal der returneres en QuestionDTO, ellers skal der returneres en QuestionStudentDTO.

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -71,51 +64,50 @@ public void create(Context ctx)
+            Question updatedEntity = dao.update(questionToUpdate);
+            ctx.status(200).json(new QuestionDTO(updatedEntity));
         }
         catch (Exception ex)
```
Tilføj catch(DaoException ex) {
logger det samme
throw new ApiException(400, "Field 'xx' is required");

MojoOno committed:
- Added functionality to getById() so that the standard output is a QuestionStudentDTO but you have the option to query output=full to get the full QuestionDTO output
- Added a catch clause for a DaoException to the update method
- Changes Long.class to Integer.class + the id is saved as an Integer as well

AndyTheDragon committed:
Refactor GenericDAO methodname getAll to getMany

AndyTheDragon committed:
Implemented overloaded getMany to allow pagination

MojoOno committed:
added manual test to demo.http

AndyTheDragon committed:
Merge remote-tracking branch 'origin/Franck/QuestionImplementation' into andy/genericdao-pagination

# Conflicts:
#	src/main/java/dat/controllers/QuestionController.java

AndyTheDragon reviewed (commented):
Ok, lad os arbejde på pagination i getAll metoden

AndyTheDragon committed:
Merge pull request #9 from AndyTheDragon/andy/genericdao-pagination

andy/genericdao-pagination

MojoOno committed:
added pagination to getAll()

MojoOno committed:
refactored all entity to question

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -71,76 +82,87 @@ public void create(Context ctx)
+        }
     }
+
+//    public void getCategory(Context ctx)
```
Disse udkommenterede kodestumper skal vi vist ikke bruge længere. Ud med dem.

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -25,44 +27,53 @@ public QuestionController(EntityManagerFactory emf)
         dao = new HotelDAO(emf);
```
Det er vist ikke den rigtige DAO at kalde her.

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -71,76 +82,87 @@ public void create(Context ctx)
-            }
-            if (incomingEntity.getAddress() != null)
+            QuestionDTO questionDTO = ctx.bodyAsClass(QuestionDTO.class);
+
```
Hvorfor ikke bruge .check( (dto)-> dto.getId() == null || dto.getId()<1)
.getOrThrow( (validator)-> new BadResponseRequest("invalid id") ?

MojoOno replied:
Den siger jeg ikke kan bruge .check på en dto men kun på id. Hvis du allerede har løst den hjemmefra, og det absolut skal være sådan (pga ensretning i koden), så må du meget gerne fjerne if statement og selv klappe den ind :)

AndyTheDragon replied:
https://javalin.io/documentation#validation
Dokumentationen siger man skal bruge ctx.bodyValidator i stedet for:
QuestionDTO questionDTO = ctx.bodyValidator(QuestionDTO.class)
.check(q -> q.getId() != null, "id must not be null")
.check(q -> q.getId() > 0, "id must be at least 1")
.getOrThrow((validator) -> new BadRequestResponse("Invalid id"));

AndyTheDragon commented on src/main/java/dat/controllers/QuestionController.java:
```diff
@@ -71,76 +82,87 @@ public void create(Context ctx)
+
+            Question questionToUpdate = dao.getById(Question.class, questionDTO.getId());
+
+            updateFieldIfNotNull(questionToUpdate::setTermDate, questionDTO.getTermDate());
```
Giver det mening med endnu en hjælpmetode 
private void updateQuestionIfNotNull(Question, QuestionsDTO) {
9x updateFieldIfNotNull
} ?

AndyTheDragon reviewed (changes requested):
Keep up the good work.

MojoOno committed:
Made first draft for test environment for the QuestionController.
Test success on getAll, get by ID and create

The rest is still work in progress.

AndyTheDragon committed:
Start på test af Questions, kopieret ind fra Mathias.

MojoOno committed:
Added constructor without localdatetime

MojoOno committed:
- Added new helper method updateQuestionIfNotNull and refactored the updateField calls down into this method to make the update method more clean.
- deleted obsolote code + whitespace

MojoOno committed:
- Added new helper method to interface

MojoOno committed:
Merge branch 'Franck/QuestionImplementation' into Franck/QuestionControllerTest

# Conflicts:
#	src/main/java/dat/controllers/QuestionController.java

MojoOno committed:
- Deleted the rogue test class

MojoOno committed:
- Moved the helper method one line up

MojoOno committed:
- Removed obsolete imports
- Refactored the two entities
- Refactored the getById() and tested for success
- Refactored create()
- Refactored update() to tage a responsebody
- - Note that the test is passing with a 200 status code, but if i use the 204 from documentation, it fails.

All tests have been tested and is passing

MojoOno committed:
Merge branch 'dev' into Franck/QuestionImplementation

# Conflicts:
#	src/main/java/dat/entities/Question.java

MojoOno committed:
Merge branch 'dev' into Franck/QuestionControllerTest

# Conflicts:
#	src/main/java/dat/entities/Question.java

MojoOno committed:
Merge branch 'Franck/QuestionControllerTest' into Franck/QuestionImplementation

MojoOno committed:
Implemented the dying wish for the code to be the same from start to bottom - (which is correct and i stand corrected)

MojoOno committed:
Implemented the dying wish for the code to be the same from start to bottom - (which is correct and i stand corrected)

MojoOno committed:
From hotel to GenericDAO

AndyTheDragon committed:
fixed hardcoded id in update test
